### DISCUSSION TITLE
Plugins Restart

### DISCUSSION THREAD

**Discussion by shyshywhy** [eclipse-theia/theia#10420] [Ideas]
Why does Plugin need to restart after the network is disconnected? It may just be a short-term network failure. If there is a way to prevent user operations after a short network disconnection, is it possible to reconnect without restarting the plugin process?

**Comment by shyshywhy** [comment]
Supplementary reason: There are many Plugins that are stateful, and startup is time-consuming, and frequent restarts lead to poor experience

**Comment by msujew** [comment]
The backend always treats a reconnecting client (one that was temporarily disconnected) in the same way as a new client. Since Theia starts a new plugin host for each new client, it will also start a new plugin host for reconnecting clients. I don't think there's currently a way around that without some refactoring of the networking code. You can create a feature-request for that, contributions are always welcome :)

**Reply by shyshywhy** [reply]
> The backend always treats a reconnecting client (one that was temporarily disconnected) in the same way as a new client. Since Theia starts a new plugin host for each new client, it will also start a new plugin host for reconnecting clients. I don't think there's currently a way around that without some refactoring of the networking code. You can create a feature-request for that, contributions are always welcome :)

My idea is to generate a session id for each browser tab to indicate identity, so that it can distinguish between reconnection and new opening (refresh is considered a new opening). After the disconnection, a dialog is used to prevent user operations, and the network request during the disconnection process is recorded through the queue, and then replayed after recovery. The default longest recovery time is 15s.
